File: code/ex858.java
class Solution {
    public int mirrorReflection(int p, int q) {
        int m = lcd(p, q) / q;
        int n = lcd(p, q) / p;
        return m % 2 == 0 ? 2 : n % 2;
    }
    
    private int lcd(int a, int b) {
        return a * b / gcd(a, b);
    }
    
    private int gcd(int a, int b) {
        if (b == 0)    return a;
        return gcd(b, a % b);
    }
}

File: code/ex858.cpp
class Solution {
public:
    int mirrorReflection(int p, int q) {
        int x = lcd(p, q) / q;
        int y = lcd(p, q) / p;
        return y % 2 == 0 ? 0 : x % 2 == 0 ? 2 : 1;
    }
    
private:
    int lcd(int p, int q) {
        return p * q / GCD(p, q);
    }
    
    int GCD(int p, int q) {
        if (q == 0)    return p;
        return GCD(q, p % q);
    }
};